In [ ]:
# env uacademia

![](https://i.ytimg.com/vi/dmrzO3T-GXQ/maxresdefault.jpg)

# Análisis de Datos Exploratorias (EDA)

## Introducción

El análisis de datos exploratorios (**EDA**) es un paso importante en el proceso de análisis de datos para comprender mejor el conjunto de datos. Al hacer EDA, podemos entender las principales características de los datos, las relaciones entre las variables y las variables que son relevantes para nuestro problema. La EDA también puede ayudar a identificar y manejar valores faltantes o duplicados, atípicos y errores en los datos.

Para este proyecto, se analizara los Goodreads Choice Awards Los Mejores Libros de 2023.

![](https://miro.medium.com/v2/resize:fit:720/format:webp/0*s8PTj_UYX2LbZpl-.jpg)

## Entender los datos

### Importar las bibliotecas necesarias

In [7]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

### Cargar el conjunto de datos en pandas dataframe

In [15]:
#Los datos están en un archivo CSV, así que se va a utilizar Pandas para leer el archivo CSV.
df = pd.read_csv('Good_Reads_Book_Awards_Crawl_2023_12_27_11_14.csv')

df.sample(5) #mostrar un set de los datos

,source_URL,Readers Choice Votes,Readers Choice Category,Title,Author,Total Avg Rating,Number of Ratings,Number of Reviews,Number of Pages,Edition,Book Description,First Published date,Kindle Version and Price,Kindle Price,About the Author
70,https://www.goodreads.com/book/show/61797464-m...,10557,Romance,Meet Me at the Lake,Carley Fortune,3.71,"158,581","15,397",336,Paperback,Fern Brookbanks has wasted far too much of her...,2-May-23,Kindle $10.99,10.99,Carley Fortune is the #1 New York Times and #1...
113,https://www.goodreads.com/book/show/61361681-c...,6588,Fantasy,Clytemnestra,Costanza Casati,4.26,"14,429","2,525",448,Hardcover,"For fans of Madeline Miller's Circe, a stunnin...",2-May-23,Kindle $8.57,8.57,Costanza Casati was born in Texas in 1995 and ...
297,https://www.goodreads.com/book/show/61089466-a...,1137,Humor,Alexandra Petri's US History: Important Americ...,Alexandra Petri,3.72,600,164,326,Hardcover,"A witty, absurdist satire of the last 500 year...",11-Apr-23,Kindle $14.43,14.43,My new book is AP's US HISTORY: Important Amer...
263,https://www.goodreads.com/book/show/61858962-v...,17446,History & Biography,Valiant Women: The Extraordinary American Serv...,Lena Andrews,4.04,640,117,351,Hardcover,A groundbreaking new history of the role of Am...,1-Aug-23,Kindle $13.99,13.99,Lena Andrews is the author of Valiant Women: T...
53,https://www.goodreads.com/book/show/63833664-a...,6741,Mystery & Thriller,After That Night,Karin Slaughter,4.36,"27,413","2,646",432,Hardcover,"After that night, nothing was ever the same ag...",22-Aug-23,Kindle $14.99,14.99,Karin Slaughter is the author of more than twe...


Entonces se va a eliminar algunas columna innecesaria del conjunto de datos, este paso es opcional, pero ya que no se va a usar esas columnas, es mejor eliminarlas para reducir el tamaño del DataFrame

In [16]:
#Las columnas no utilizadas son source.URL, Book Description y About the Author
df.drop(['source_URL','Book Description','About the Author'],axis=1, inplace=True)

### Comprobación del DataFrame

Ahora se va a comprobar los tipos de datos para cada columna y comprobar el resumen de las columnas numéricas para que se pueda determinar la próxima acción.

In [17]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Readers Choice Votes      299 non-null    int64  
 1   Readers Choice Category   299 non-null    str    
 2   Title                     299 non-null    str    
 3   Author                    299 non-null    str    
 4   Total Avg Rating          299 non-null    float64
 5   Number of Ratings         299 non-null    str    
 6   Number of Reviews         299 non-null    str    
 7   Number of Pages           299 non-null    int64  
 8   Edition                   299 non-null    str    
 9   First Published date      299 non-null    str    
 10  Kindle Version and Price  299 non-null    str    
 11  Kindle Price              299 non-null    float64
dtypes: float64(2), int64(2), str(8)
memory usage: 28.2 KB


Según **.info()**, el conjunto de datos se ve bien sin ningún valor faltante. También da cierta información, como la forma de nuestro conjunto de datos (12 columnas y 299 filas) y los tipos de datos para cada columna.

In [18]:
df.describe()

,Readers Choice Votes,Total Avg Rating,Number of Pages,Kindle Price
count,299.000000,299.000000,299.000000,299.000000
mean,19662.919732,4.063512,392.394649,11.327926
std,32406.337679,0.252188,123.698044,5.351909
min,935.000000,3.520000,192.000000,0.000000
25%,3591.000000,3.855000,320.000000,9.990000
50%,10099.000000,4.060000,368.000000,13.990000
75%,24714.000000,4.250000,432.000000,14.990000
max,397565.000000,4.750000,1242.000000,17.990000


El método **.describe()**  da estadísticas resumidas para columnas numéricas del DataFrame. Tales como el promedio, media, desviación estándar, valores mínimos y máximos para cada columna numérica.

### Cambio de Int/Float y asignación de tipos de datos

Una vez que se ha identificado el tamaño y los tipos numericos de los datos, se puede proceder a determinar los pasos siguientes delo análsis de losdatos. Por **.info()**, conocemos el tamaño de los datos, que es de **28.2 KB** y los tipos de datos para cada columna. El método **.describe()** nos mostró las estadísticas de las columnas numéricas, como el valor mínimo y máximo de cada columna, así como el promedio.

Con este resultado, se puede ver que todavía faltan algunas columnas, como el número de calificaciones y el número de críticas, que se supone que son columnas numéricas. Resulta que esas columnas usan comas “,”como separador decimal. Otras columnas como *Readers Choice Votes* no tienen separadores de miles, ya que se almacenaron como formato general. La razón para escribir números con un separador y en vez usar el número completo es porque es era opción más segura, ya que los números podrían ser identificadores, o algún otro contador de número. Así que, para hacer eso, se necesita quitar la coma.

In [19]:
numeric_columns = ['Number of Ratings','Number of Reviews']

#Elimina las comas de las columnas y convierte los datos a Int32
for column in numeric_columns:
    df[column] = df[column].replace(',', '', regex=True).astype('int32')

In [20]:
df.describe()

,Readers Choice Votes,Total Avg Rating,Number of Ratings,Number of Reviews,Number of Pages,Kindle Price
count,299.000000,299.000000,299.000000,299.000000,299.000000,299.000000
mean,19662.919732,4.063512,38748.351171,5241.705686,392.394649,11.327926
std,32406.337679,0.252188,79586.655592,10800.069660,123.698044,5.351909
min,935.000000,3.520000,228.000000,71.000000,192.000000,0.000000
25%,3591.000000,3.855000,5516.500000,952.000000,320.000000,9.990000
50%,10099.000000,4.060000,14519.000000,2469.000000,368.000000,13.990000
75%,24714.000000,4.250000,37143.000000,5109.500000,432.000000,14.990000
max,397565.000000,4.750000,930928.000000,141668.000000,1242.000000,17.990000


Este método eliminará todas las **comas** de esas columnas. Se puede dejar por defecto como **.astype('int32')** porque pandas asignará automáticamente el tipo de datos en **int64**. Los números 32 y 64 representan la cantidad de memoria o cuántos bits por fila va a almacenar estos valores.

In [21]:
df.describe ()

,Readers Choice Votes,Total Avg Rating,Number of Ratings,Number of Reviews,Number of Pages,Kindle Price
count,299.000000,299.000000,299.000000,299.000000,299.000000,299.000000
mean,19662.919732,4.063512,38748.351171,5241.705686,392.394649,11.327926
std,32406.337679,0.252188,79586.655592,10800.069660,123.698044,5.351909
min,935.000000,3.520000,228.000000,71.000000,192.000000,0.000000
25%,3591.000000,3.855000,5516.500000,952.000000,320.000000,9.990000
50%,10099.000000,4.060000,14519.000000,2469.000000,368.000000,13.990000
75%,24714.000000,4.250000,37143.000000,5109.500000,432.000000,14.990000
max,397565.000000,4.750000,930928.000000,141668.000000,1242.000000,17.990000


Se visualiza el valor mínimo y máximo de cada columna numérica. Por ejemplo de la columna *Readers Choice Votes*, el valor más pequeño es el **935** y el más alto es **397565**.

Conociendo el rango de valores, ahora se puede determinar los bits para almacenar estos valores. 
Para referencias:

- **Int8** Las variables pueden contener valores que van de -128 a 127.
- **Int16** Las variables pueden contener valores que oscilan entre 3.778 y 32.767.
- **Int32** Las variables pueden contener valores que oscilan entre -2,147,483,648 y 2.147.483.647.
- **Int64**Las variables pueden almacenar valores que oscilan entre -9,223,372.0366,854,775,808 a 9.223,372,0366,854,775,807.

El **Int32** es la opción perfecta porque se ajusta al rango de valor. Se puede usar Int64, pero no es apropiado porque utilizaría más memoria y hace que el DataFrame sea menos eficiente.

Para decimales (float), es un poco diferente porque realmente afecta a cuántos decimales se puede almacenar en  loa datos. **Float16** almacena **4** dígitos decimales, **Float32** almacena **8** dígitos decimales, y **Float64** almacena **16** dígitos decimales. La mejor opción es usar Float16 porque realmente se necesita usar muchos dígitos decimales en el DataFrame, pero todavía se quiere mantener los valores iguales al original.

Ahora que se conoce el rango de valor para cada columna, se asignara a cada una de esas columnas los tipos de datos adecuados.

También hay algunas columnas que almacenan valores de texto. Se puede asignar los tipos de datos de esas columnas como **string** o **category**. A partir de la documentación de pandas, el tipo de datos categórico es útil si una variable de texto contiene en sólo unos pocos valores diferentes, por ejemplo, género, clase social, tipo de sangre, afiliación al país, etc. Por esa definición, la categoría es la mejor elección de las columnas de textp para usar como tipo de datos categórico.

In [22]:
#Convierte el resto de las columnas al correcto tipo de datos
convert_dict = {'Readers Choice Votes': 'int32',
                'Readers Choice Category': 'category',
                'Title': 'string',
                'Author': 'string',
                'Total Avg Rating': 'float16',
                'Number of Pages': 'int16',
                'Edition': 'category',
                'First Published date': 'datetime64[ns]',
                'Kindle Price': 'float16'}
df = df.astype(convert_dict)

In [23]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   Readers Choice Votes      299 non-null    int32         
 1   Readers Choice Category   299 non-null    category      
 2   Title                     299 non-null    string        
 3   Author                    299 non-null    string        
 4   Total Avg Rating          299 non-null    float16       
 5   Number of Ratings         299 non-null    int32         
 6   Number of Reviews         299 non-null    int32         
 7   Number of Pages           299 non-null    int16         
 8   Edition                   299 non-null    category      
 9   First Published date      299 non-null    datetime64[ns]
 10  Kindle Version and Price  299 non-null    str           
 11  Kindle Price              299 non-null    float16       
dtypes: category(2), datetime64[ns](1)

Para la columna **Kindle Version and Price**, se eliminara el precio, ya que la otra columna **Kindle Price** contiene los datos de los precios

In [24]:
#Separar el precio del texto y crear una nueva columna
df['Kindle Version'] = df['Kindle Version and Price'].str.extract('([a-zA-Z ]+)', expand=False).str.strip()

#Cambiar el tipo de dato a la nueva columna
df['Kindle Version'] = df['Kindle Version'].astype('category')

#Eliminar la columna original
df = df.drop('Kindle Version and Price', axis=1)

Revisemos una vez más nuestro dataframe para ver los tipos de datos como han cambiado:

In [25]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 12 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   Readers Choice Votes     299 non-null    int32         
 1   Readers Choice Category  299 non-null    category      
 2   Title                    299 non-null    string        
 3   Author                   299 non-null    string        
 4   Total Avg Rating         299 non-null    float16       
 5   Number of Ratings        299 non-null    int32         
 6   Number of Reviews        299 non-null    int32         
 7   Number of Pages          299 non-null    int16         
 8   Edition                  299 non-null    category      
 9   First Published date     299 non-null    datetime64[ns]
 10  Kindle Price             299 non-null    float16       
 11  Kindle Version           299 non-null    category      
dtypes: category(3), datetime64[ns](1), float16(2), 

In [26]:
df.describe()

,Readers Choice Votes,Total Avg Rating,Number of Ratings,Number of Reviews,Number of Pages,First Published date,Kindle Price
count,299.000000,299.000000,299.000000,299.000000,299.000000,299,299.000000
mean,19662.919732,4.062500,38748.351171,5241.705686,392.394649,2023-05-19 09:57:11.438127104,11.328125
min,935.000000,3.519531,228.000000,71.000000,192.000000,2019-11-27 00:00:00,0.000000
25%,3591.000000,3.854492,5516.500000,952.000000,320.000000,2023-02-28 12:00:00,9.992188
50%,10099.000000,4.058594,14519.000000,2469.000000,368.000000,2023-05-16 00:00:00,13.992188
75%,24714.000000,4.250000,37143.000000,5109.500000,432.000000,2023-08-29 00:00:00,14.992188
max,397565.000000,4.750000,930928.000000,141668.000000,1242.000000,2023-11-13 00:00:00,17.984375
std,32406.337679,0.252197,79586.655592,10800.069660,123.698044,NaN,5.351562


In [27]:
df.sample(10)

,Readers Choice Votes,Readers Choice Category,Title,Author,Total Avg Rating,Number of Ratings,Number of Reviews,Number of Pages,Edition,First Published date,Kindle Price,Kindle Version
71,10099,Romance,Pucking Around,Emily Rath,3.949219,78119,8274,756,Kindle Edition,2023-03-18,0.000000,Kindle Unlimited
185,22755,Young Adult Fiction,Highly Suspicious and Unfairly Cute,Talia Hibbert,3.759766,24499,5681,336,ebook,2023-01-03,8.992188,Kindle
100,75800,Fantasy,Hell Bent,Leigh Bardugo,4.160156,94667,13508,481,Hardcover,2023-01-10,14.992188,Kindle
97,1738,Romantasy,A Queen of Thieves & Chaos,K.A. Tucker,4.269531,2846,465,606,Kindle Edition,2023-09-05,5.988281,Kindle
205,20045,Debut Novel,Looking for Jane,Heather Marshall,4.378906,29879,3669,400,Hardcover,2022-03-01,12.992188,Kindle
29,19468,Historical Fiction,Looking for Jane,Heather Marshall,4.378906,29873,3669,400,Hardcover,2022-03-01,12.992188,Kindle
249,4034,Memoir & Autobiography,Thicker than Water: A Memoir,Kerry Washington,3.919922,6981,718,320,Hardcover,2023-09-26,14.992188,Kindle
81,33665,Romantasy,Assistant to the Villain,Hannah Nicole Maehrer,3.939453,50630,10209,342,Paperback,2023-08-29,12.992188,Kindle
182,57736,Young Adult Fiction,Five Survive,Holly Jackson,4.070312,98441,17770,388,Hardcover,2022-11-29,10.992188,Kindle
194,3794,Young Adult Fiction,What Happens After Midnight,K.L. Walther,3.779297,11048,2245,384,Paperback,2023-06-27,7.390625,Kindle


Como se observa en los resultados, se puede **reducir el tamaño** de memoria del DataFrame cambiando los tipos de datos y reduciendo enteros y decimales. Hay que tener en cuenta que estos pasos son **opcionales**, pero puede ser **muy útil si se esta trabajando con un gran conjunto de datos**.